### PR TITLE
JWT Token Generation with Private and Public Key

### PR DESCRIPTION
**Pull request by susilnem**
Addresses 
 - https://github.com/IFRCGo/go-api/issues/1982

## Changes
- Create a Model for ExternalToken Generation
- Create test cases for ExternalToken
- Add accepted_montandon_license_terms field
- Api for montadon license terms 

## Before  merge!
Please note that it requires new env variables. You can generate the values for the token using the bash script

```bash
#!/usr/bin/env bash

mkdir -p keys/

function gen_key {
    # EdDSA
    # openssl genpkey -algorithm ed25519 -out keys/$1
    # openssl pkey -in keys/$1 -pubout > keys/$1.pub

    # ES256
    openssl ecparam -name prime256v1 -genkey -noout -out keys/$1
    openssl ec -in keys/$1 -pubout > keys/$1.pub
}

# Used
gen_key jwtkey-01
```

## Env variable to set 
```bash
 JWT_PRIVATE_KEY
 JWT_PUBLIC_KEY
 JWT_EXPIRE_TIMESTAMP_DAYS
```
## Checklist
Things that should succeed before merging.
- [ ] Updated/ran unit tests
- [ ] Updated CHANGELOG.md

## Release

If there is a version update, make sure to tag the repository with the latest version.

### CODE REVIEW

File: registrations/models.py [hunk 1]
```diff
@@ -71,3 +71,24 @@ class Meta:
...
+    """ External token for user """
+
+    user = models.ForeignKey(
+        settings.AUTH_USER_MODEL,
+        verbose_name=_('user'),
+        on_delete=models.CASCADE,
+    )
+
+    created_at = models.DateTimeField(verbose_name=_('created at'), auto_now_add=True)
+    token = models.TextField(verbose_name=_('token'), editable=False)
```
Comment (thenav56) [inline registrations/models.py:85]: We only store jti instead of whole token

File: registrations/models.py [hunk 2]
```diff
@@ -71,3 +71,24 @@ class Meta:
...
+    expire_timestamp = models.DateTimeField(verbose_name=_('expire timestap'), null=True, blank=True)
+    is_disabled = models.BooleanField(verbose_name=_('is disabled?'), default=False)
```
Comment (thenav56) [inline registrations/models.py:87]: For now, let's comment it out and add a note, as we haven't implemented blacklist in this phase.

File: registrations/models.py [hunk 3]
```diff
@@ -71,3 +71,24 @@ class Meta:
...
+
+    class Meta:
+        verbose_name = _('User External Token')
+        verbose_name_plural = _('User External Tokens')
+
+    def __str__(self):
+        return f'{self.user.username} ({self.created_at.strftime("%Y-%m-%d")})'
```
Comment (thenav56) [inline registrations/models.py:94]:
```suggestion
        return f'{self.user_id} ({self.created_at.strftime("%Y-%m-%d")})'
```

File: registrations/serializers.py [hunk 4]
```diff
@@ -158,3 +163,37 @@ def save(self):
...
+class UserExternalTokenSerializer(serializers.ModelSerializer):
+    user_details = UserNameSerializer(source="user", read_only=True)
+    class Meta:
+        model = UserExternalToken
+        fields = 'id', 'user_details', 'token', 'expire_timestamp'
+
+    def validate_expire_timestamp(self, data):
+        if data < timezone.now():
+            raise serializers.ValidationError('Expire timestamp must be in the future.')
+        return data
```
Comment (thenav56) [inline registrations/serializers.py:176]: Let's add a max limit as well

```suggestion
    def validate_expire_timestamp(self, date):
        if date < timezone.now():
            raise serializers.ValidationError('Expire timestamp must be in the future.')
        return date
```

File: registrations/serializers.py [hunk 5]
```diff
@@ -158,3 +163,37 @@ def save(self):
...
+
+    def create(self, validated_data):
+        validated_data['user'] = self.context['request'].user
+
+        if not validated_data.get('expire_timestamp'):
+            validated_data['expire_timestamp'] = timezone.now() + timedelta(days=7)
```
Comment (thenav56) [inline registrations/serializers.py:182]: Default will be a bit more than that. We need to discuss this.

File: registrations/serializers.py [hunk 6]
```diff
@@ -158,3 +163,37 @@ def save(self):
...
+        
+        # Check if private and public key exists
+        if not(settings.JWT_PRIVATE_KEY and settings.JWT_PUBLIC_KEY):
+            raise serializers.ValidationError('Please Provide private and public key.')
```
Comment (thenav56) [inline registrations/serializers.py:186]:
```suggestion
            raise serializers.ValidationError('Please contact system adminstrators to configurate private and public key.')
```

File: registrations/serializers.py [hunk 7]
```diff
@@ -158,3 +163,37 @@ def save(self):
...
+        
+        payload = {
+            'user': self.context['request'].user.id,
+            'exp': validated_data['expire_timestamp']
+        }
+
+        validated_data['token'] = jwt_encode_handler(payload)
```
Comment (thenav56) [inline registrations/serializers.py:193]: We don't store the token to the database, we only need to send it back when creating a new token.
NOTE: We store jit instead

File: registrations/utils.py [hunk 8]
```diff
@@ -45,3 +52,18 @@ def getRegionalAdmins(userId):
 
     admins = UserRegion.objects.filter(region_id=regionId).values_list('user__email', flat=True)
     return admins
+
+
+def jwt_encode_handler(payload):
+    return jwt.encode(
+        payload,
+        settings.JWT_PRIVATE_KEY,
+        algorithm='RS256',
```
Comment (thenav56) [inline registrations/utils.py:61]: We are using `ES256`

File: registrations/views.py [hunk 9]
```diff
@@ -134,3 +138,19 @@ def get(self, request):
             )
             pending_user.delete()
             return HttpResponse(render_to_string('registration/validation-success.html'))
+
+class UserExternalTokenViewset(viewsets.ModelViewSet):
+    queryset = UserExternalToken.objects.exclude(is_disabled=True)
```
Comment (thenav56) [inline registrations/views.py:143]: We can show all the user tokens

```suggestion
    queryset = UserExternalToken.objects.exclude(is_disabled=True)
```

File: registrations/views.py [hunk 10]
```diff
@@ -134,3 +138,19 @@ def get(self, request):
...
+    def get_queryset(self):
+        return super().get_queryset().filter(user=self.request.user)
+    
+    def destroy(self, request, *args, **kwargs):
+        instance = self.get_object()
+        instance.is_disabled = True
+        instance.save()
+        return Response(
+            status=status.HTTP_204_NO_CONTENT
+        )
```
Comment (thenav56) [inline registrations/views.py:156]: We don't need to delete, instead just disable. Which will be used when we implement blacklisting.

File: registrations/models.py [hunk 11]
```diff
@@ -71,3 +71,24 @@ class Meta:
 
     def __str__(self):
         return self.domain_name
+
+class UserExternalToken(models.Model):
+    """ External token for user """
+
```
Comment (thenav56) [inline registrations/models.py:77]: Let's add a title

**Review by thenav56 — CHANGES_REQUESTED** [review]

File: api/drf_views.py [hunk 12]
```diff
@@ -875,7 +875,19 @@ def get_queryset(self):
...
         return Response(self.get_serializer_class()(request.user).data)
-
+    
+    @action(
+        detail=True,
+        methods=['post'],
+        url_path="accepted_license_terms",
+    )
+    def accepted_license_terms(self, request, *args, **kwargs):
+        user = self.get_object()
```
Comment (thenav56) [inline api/drf_views.py:885]:
```suggestion
        user = request.user
```

File: main/settings.py [hunk 13]
```diff
@@ -96,6 +96,10 @@
     DJANGO_READ_ONLY=(bool, False),
     # Misc
     DISABLE_API_CACHE=(bool, False),
+    #jwt private and public key
```
Comment (thenav56) [inline main/settings.py:99]:
```suggestion
    # jwt private and public key
```

File: registrations/models.py [hunk 14]
```diff
@@ -71,3 +71,35 @@ class Meta:
...
+        verbose_name=_('user'),
+        on_delete=models.CASCADE,
+    )
+
+    jti = models.CharField(
+        editable=False,
+        max_length=32,
+        verbose_name=_('jti'),
+        help_text=_('Unique identifier for the token')
+    )
```
Comment (thenav56) [inline registrations/models.py:94]: Let's use uuid field instead of charfield

File: registrations/serializers.py [hunk 15]
```diff
@@ -158,3 +162,47 @@ def save(self):
...
+        now = timezone.now()
+        if date < now:
+            raise serializers.ValidationError('Expire timestamp must be in the future.')
+        elif date > now + timedelta(days=settings.JWT_EXPIRE_TIMESTAMP_DAYS):
+            raise serializers.ValidationError('Expire timestamp must be less than 1 year.')
+        return date
+
+    def create(self, validated_data):
+        validated_data['user'] = user = self.context['request'].user
+        validated_data['jti'] = get_random_string(length=32)
```
Comment (thenav56) [inline registrations/serializers.py:183]: Let's use uuid4

File: registrations/serializers.py [hunk 16]
```diff
@@ -158,3 +162,47 @@ def save(self):
...
+
+        if not user.profile.accepted_montandon_license_terms:
+            raise serializers.ValidationError('User must accept Montandon license terms.')
```
Comment (thenav56) [inline registrations/serializers.py:186]: For now we skip this..  so let's comment it out.

File: registrations/serializers.py [hunk 17]
```diff
@@ -158,3 +162,47 @@ def save(self):
...
+        # Check if private and public key exists
+        if not(settings.JWT_PRIVATE_KEY and settings.JWT_PUBLIC_KEY):
+            raise serializers.ValidationError('Please contact system adminstrators to configurate private and public key.')
+        
+        payload = {
+            'jti': validated_data['jti'],
+            'userId': user.id,
+            'exp': validated_data['expire_timestamp'],
+            "inMovement": True,
+        }
```
Comment (thenav56) [inline registrations/serializers.py:200]: Let's move this section to model

File: registrations/serializers.py [hunk 18]
```diff
@@ -158,3 +162,47 @@ def save(self):
...
+
+        UserExternalToken.objects.create(**validated_data)
```
Comment (thenav56) [inline registrations/serializers.py:202]:
```suggestion
        instance = super().create(validated_data)
```

File: registrations/models.py (hunk 3 above)
Comment (thenav56) [reply registrations/models.py]: Let's add an expiry information there as well.

File: registrations/serializers.py [hunk 19]
```diff
@@ -158,3 +162,47 @@ def save(self):
...
+    class Meta:
+        model = UserExternalToken
+        fields = ['title', 'token', 'expire_timestamp']
+
+    def validate_expire_timestamp(self, date):
+        now = timezone.now()
+        if date < now:
+            raise serializers.ValidationError('Expire timestamp must be in the future.')
+        elif date > now + timedelta(days=settings.JWT_EXPIRE_TIMESTAMP_DAYS):
+            raise serializers.ValidationError('Expire timestamp must be less than 1 year.')
```
Comment (thenav56) [inline registrations/serializers.py:178]:
```suggestion
            raise serializers.ValidationError('Expire timestamp must be less than {settings.____} days.')
```

**Review by thenav56 — CHANGES_REQUESTED** [review]

File: registrations/test_views.py [hunk 20]
```diff
@@ -164,3 +170,52 @@ def test_user_registration(self, send_notification_create):
...
+
+        data = {
+            "title": "ok"
+        }
+
+        with override_settings(
+            JWT_PRIVATE_KEY=self.JWT_PRIVATE_KEY,
+            JWT_PUBLIC_KEY=self.JWT_PUBLIC_KEY,
+        ):
+            response = self.client.post('/api/v2/external-token/', data, format='json')
```
Comment (thenav56) [inline registrations/test_views.py:215]: Let's also check if response have the newly generated token and another basic GET check as well.

**Review by thenav56 — CHANGES_REQUESTED** [review]
Last minor change

File: api/drf_views.py [hunk 21]
```diff
@@ -875,7 +875,19 @@ def get_queryset(self):
...
+        methods=['post'],
+        url_path="accepted_license_terms",
+    )
+    def accepted_license_terms(self, request, *args, **kwargs):
+        user = request.user
+        if user.profile.accepted_montandon_license_terms is True:
+            raise serializers.ValidationError("User has already accepted the license terms")
+        user.profile.accepted_montandon_license_terms = True
+        user.profile.save(update_fields=["accepted_montandon_license_terms"])
+        return Response(self.get_serializer_class()(user).data)
```
Comment (k9845) [inline api/drf_views.py:890]: @susilnem Let's also define the schema type

File: registrations/models.py [hunk 22]
```diff
@@ -71,3 +72,44 @@ class Meta:
...
+
+    jti = models.UUIDField(
+        default=uuid.uuid4,
+        editable=False,
+        unique=True,
+        help_text=_('Unique identifier for the token')
+    )
+
+    created_at = models.DateTimeField(verbose_name=_('created at'), auto_now_add=True)
+    expire_timestamp = models.DateTimeField(verbose_name=_('expire timestamp'), null=True, blank=True)
```
Comment (k9845) [inline registrations/models.py:98]: Remove null and blank.

File: registrations/models.py [hunk 23]
```diff
@@ -71,3 +72,44 @@ class Meta:
...
+        verbose_name = _('User External Token')
+        verbose_name_plural = _('User External Tokens')
+
+    def __str__(self):
+        return f'{self.title}-{self.expire_timestamp.strftime("%Y-%m-%d")}'
+    
+    def get_payload(self) -> dict:
+        return {
+            'jti': str(self.jti),
+            'userId': self.user.id,
```
Comment (k9845) [inline registrations/models.py:112]:
```suggestion
            'userId': self.user_id,
```

File: registrations/serializers.py [hunk 24]
```diff
@@ -158,3 +162,40 @@ def save(self):
                 email
             )
         )
+
+class UserExternalTokenSerializer(serializers.ModelSerializer):
+    token = serializers.CharField(read_only=True)
+    expire_timestamp = serializers.DateTimeField(required=False)
+    class Meta:
```
Comment (k9845) [inline registrations/serializers.py:169]:
```suggestion
    expire_timestamp = serializers.DateTimeField(required=False)

    class Meta:
```

File: registrations/test_views.py [hunk 25]
```diff
@@ -9,11 +9,17 @@
 from unittest import mock
 
 from rest_framework.test import APITestCase
+from cryptography.hazmat.backends import default_backend
+from cryptography.hazmat.primitives import serialization
+from cryptography.hazmat.primitives.asymmetric import ec
+from main.test_case import APITestCase as Apitestcase
```
Comment (k9845) [inline registrations/test_views.py:15]:
```suggestion
from main.test_case import APITestCase as GoAPITestCase
```

File: registrations/views.py [hunk 26]
```diff
@@ -6,13 +6,18 @@
 from django.conf import settings
 from django.http import HttpResponse
 from django.template.loader import render_to_string
+from registrations.serializers import UserExternalTokenSerializer
```
Comment (k9845) [inline registrations/views.py:9]: Move this before models import

**Review by k9845 — CHANGES_REQUESTED** [review]

**Comment by frozenhelium** [comment]
@susilnem Let's also capture `created_at` and also add it to the `GET` endpoint.
CC: @thenav56

**Comment by frozenhelium** [comment]
Hi @szabozoltan69, could you please merge and deploy it to staging. 

## Before you merge!
Please note that it requires new env variables. You can generate the values for the token using the bash script

```bash
#!/usr/bin/env bash

mkdir -p keys/

function gen_key {
    # EdDSA
    # openssl genpkey -algorithm ed25519 -out keys/$1
    # openssl pkey -in keys/$1 -pubout > keys/$1.pub

    # ES256
    openssl ecparam -name prime256v1 -genkey -noout -out keys/$1
    openssl ec -in keys/$1 -pubout > keys/$1.pub
}

# Used
gen_key jwtkey-01
```

**Review by frozenhelium — APPROVED** [review]

**Review by k9845 — APPROVED** [review]

**Pull request merged**